gateway/handlers/logs.go

```diff
@@ -0,0 +1,133 @@
+package handlers
+
+import (
+	"context"
+	"io"
+	"log"
+	"net"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+	"time"
+)
+
+const crlf = "\r\n"
+const upstreamLogsEndpoint = "/system/logs"
+
+// NewLogHandlerFunc creates and http HandlerFunc from the supplied log Requestor.
+func NewLogHandlerFunc(logProvider url.URL) http.HandlerFunc {
+
+	writeRequestURI := false
+	if _, exists := os.LookupEnv("write_request_uri"); exists {
+		writeRequestURI = exists
+	}
+
+	upstreamLogProviderBase := strings.TrimSuffix(logProvider.String(), "/")
+
+	return func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+		if r.Body != nil {
+			defer r.Body.Close()
+		}
+
+		logRequest := buildUpstreamRequest(r, upstreamLogProviderBase, upstreamLogsEndpoint)
+		if logRequest.Body != nil {
+			defer logRequest.Body.Close()
+		}
+
+		hijacker, ok := w.(http.Hijacker)
+		if !ok {
+			log.Println("LogProxy: response is not a Hijacker, required for streaming response")
+			http.NotFound(w, r)
+			return
+		}
+
+		clientConn, buf, err := hijacker.Hijack()
+		if err != nil {
+			log.Println("LogProxy: failed to hijack connection for streaming response")
+			return
+		}
+
+		defer clientConn.Close()
+		clientConn.SetWriteDeadline(time.Time{}) // allow arbitrary time between log writes
```

Comment:
Self document that `time.Time{}` (empty time value) means there is no `WriteDeadline`.

Reply:
Doc added